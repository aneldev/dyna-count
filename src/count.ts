export const countFromTo = (from: number, to: number, step = 1) => {
  return {
    for: (cb: (index: number, cbBreak: () => void) => void): void => {
      for (let index = from; index < to; index += step) {
        let shouldBreak = false;
        const cbBreak = () => shouldBreak = true;
        cb(index, cbBreak);
        if (shouldBreak) break;
      }
    },
    map: <T = any>(cb: (index: number) => T): T[] => {
      const output: T[] = [];
      countFromTo(from, to).for(index => output.push(cb(index)));
      return output;
    },
    reduce: <T = any>(cb: (acc: T, index: number) => T, initialValue: T): T => {
      let output: T = initialValue;
      countFromTo(from, to).for(index => output = cb(output, index));
      return output;
    },
  };
};

export const count = (countTo: number, step = 1) => countFromTo(0, countTo, step);
