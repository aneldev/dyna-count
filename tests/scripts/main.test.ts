import "jest";
import { count } from "../../src";
import { countFromTo } from "../../src/count";

jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;

describe('count', () => {

  test('count.for', () => {
    const collected: number[] = [];
    count(20).for(index => collected.push(index));
    expect(collected).toMatchSnapshot();
  });

  test('count.for with step', () => {
    const collected: number[] = [];
    count(10, 2).for(index => collected.push(index));
    expect(collected).toMatchSnapshot();
  });

  test('count.for with break', () => {
    const collected: number[] = [];
    count(20).for((index, breakIt) => {
      collected.push(index);
      if (index === 9) breakIt();
    });
    expect(collected.length).toBe(10);
  });

  test('count.map', () => {
    const mapped = count(20).map<number>(index => 2000 + index);
    expect(mapped).toMatchSnapshot();
  });

  test('countFromTo.map', () => {
    const mapped = countFromTo(10, 20).map<number>(index => 2000 + index);
    expect(mapped).toMatchSnapshot();
  });

  test('reduce', () => {
    const reduced = count(10).reduce<{ [key: string]: string }>((acc, index) => {
      acc[`key-${index}`] = `Key is ${index}`;
      return acc;
    }, {});
    expect(reduced).toMatchSnapshot();
  });
});
