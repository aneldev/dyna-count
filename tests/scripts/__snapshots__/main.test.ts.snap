// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`count count.for 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
]
`;

exports[`count count.for with step 1`] = `
Array [
  0,
  2,
  4,
  6,
  8,
]
`;

exports[`count count.map 1`] = `
Array [
  2000,
  2001,
  2002,
  2003,
  2004,
  2005,
  2006,
  2007,
  2008,
  2009,
  2010,
  2011,
  2012,
  2013,
  2014,
  2015,
  2016,
  2017,
  2018,
  2019,
]
`;

exports[`count countFromTo.map 1`] = `
Array [
  2010,
  2011,
  2012,
  2013,
  2014,
  2015,
  2016,
  2017,
  2018,
  2019,
]
`;

exports[`count reduce 1`] = `
Object {
  "key-0": "Key is 0",
  "key-1": "Key is 1",
  "key-2": "Key is 2",
  "key-3": "Key is 3",
  "key-4": "Key is 4",
  "key-5": "Key is 5",
  "key-6": "Key is 6",
  "key-7": "Key is 7",
  "key-8": "Key is 8",
  "key-9": "Key is 9",
}
`;
